File: scheduled-job-runs-twice/python/find_duplicate_ticks.py
"""Find Medusa v2 scheduled job ticks that fired more than once, because
more than one process is running in shared or worker WORKER_MODE against
the same database, with no distributed lock coordinating them.

This is an infrastructure and config defect, not a data problem. It only
reports the duplicate ticks, it never resends a suppressed side effect
and never deletes a workflow_execution row. Safe to run again and again.

Guide: https://www.allanninal.dev/medusa/scheduled-job-runs-twice/
"""
import os
import logging
from datetime import datetime, timedelta

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("find_duplicate_ticks")

BASE_URL = os.environ.get("MEDUSA_BACKEND_URL", "http://localhost:9000")
EMAIL = os.environ.get("MEDUSA_ADMIN_EMAIL", "admin@example.com")
PASSWORD = os.environ.get("MEDUSA_ADMIN_PASSWORD", "supersecret")
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"
JOB_WORKFLOW_ID = os.environ.get("JOB_WORKFLOW_ID", "job-name")
JOB_CRON = os.environ.get("JOB_CRON", "*/15 * * * *")
BUCKET_TOLERANCE_MS = float(os.environ.get("BUCKET_TOLERANCE_MS", "5000"))

EXECUTION_FIELDS = "id,transaction_id,workflow_id,created_at,state"


def get_token():
    r = requests.post(
        f"{BASE_URL}/auth/user/emailpass",
        json={"email": EMAIL, "password": PASSWORD},
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["token"]


def list_workflow_executions(token, workflow_id, limit=200):
    headers = {"Authorization": f"Bearer {token}"}
    r = requests.get(
        f"{BASE_URL}/admin/workflows-executions",
        params={"workflow_id": workflow_id, "fields": EXECUTION_FIELDS, "limit": limit,
                "order": "created_at"},
        headers=headers,
        timeout=30,
    )
    r.raise_for_status()
    return r.json()["workflow_executions"]


def _parse_field(field, lo, hi):
    """Parse one cron field (*, N, N-M, N,M, */S) into a set of allowed ints."""
    if field == "*":
        return set(range(lo, hi + 1))
    values = set()
    for part in field.split(","):
        if part.startswith("*/"):
            values.update(range(lo, hi + 1, int(part[2:])))
        elif "-" in part:
            a, b = part.split("-")
            values.update(range(int(a), int(b) + 1))
        else:
            values.add(int(part))
    return values


def _parse_cron(cron_expression):
    minute, hour, dom, month, dow = cron_expression.strip().split()
    return {
        "minute": _parse_field(minute, 0, 59),
        "hour": _parse_field(hour, 0, 23),
        "dom": _parse_field(dom, 1, 31),
        "month": _parse_field(month, 1, 12),
        "dow": _parse_field(dow, 0, 6),
    }


def _matches(dt, spec):
    if dt.minute not in spec["minute"] or dt.hour not in spec["hour"] or dt.month not in spec["month"]:
        return False
    dom_ok = dt.day in spec["dom"]
    dow_ok = (dt.weekday() + 1) % 7 in spec["dow"]
    if spec["dom"] != set(range(1, 32)) and spec["dow"] != set(range(0, 7)):
        return dom_ok or dow_ok
    return dom_ok and dow_ok


def nearest_tick_boundary(cron_expression, at, search_minutes=1440):
    """Minute-aligned tick boundary matching the cron spec closest to `at`."""
    spec = _parse_cron(cron_expression)
    base = at.replace(second=0, microsecond=0)
    if _matches(base, spec):
        return base
    for offset in range(1, search_minutes + 1):
        earlier = base - timedelta(minutes=offset)
        if _matches(earlier, spec):
            return earlier
        later = base + timedelta(minutes=offset)
        if _matches(later, spec):
            return later
    raise RuntimeError("No matching tick boundary found within search window")


def find_duplicate_ticks(executions, cron_schedule, bucket_tolerance_ms=5000):
    """Pure: no I/O, no clock access. executions is a list of
    {"workflow_id", "transaction_id", "created_at"} dicts already fetched."""
    by_workflow = {}
    for execution in executions:
        by_workflow.setdefault(execution["workflow_id"], []).append(execution)

    duplicates = []
    for workflow_id, rows in by_workflow.items():
        buckets = {}
        for row in rows:
            created_at = datetime.fromisoformat(row["created_at"].replace("Z", "+00:00"))
            tick = nearest_tick_boundary(cron_schedule, created_at)
            delta_ms = abs((created_at - tick).total_seconds() * 1000)
            if delta_ms > bucket_tolerance_ms:
                continue
            key = tick.isoformat()
            buckets.setdefault(key, set()).add(row["transaction_id"])

        for tick_bucket, tx_ids in buckets.items():
            if len(tx_ids) > 1:
                duplicates.append({
                    "tickBucket": tick_bucket,
                    "transactionIds": sorted(tx_ids),
                })

    duplicates.sort(key=lambda d: d["tickBucket"])
    return duplicates


def write_audit_report(job_name, duplicates):
    """The only write this script does: an audit log line per duplicate tick.
    Never resends a suppressed side effect, never deletes an execution row."""
    for item in duplicates:
        log.warning(
            "DUPLICATE TICK job=%s tick=%s transaction_ids=%s inferred_replicas=%d",
            job_name, item["tickBucket"], item["transactionIds"], len(item["transactionIds"]),
        )


def run():
    token = get_token()
    executions = list_workflow_executions(token, JOB_WORKFLOW_ID)
    duplicates = find_duplicate_ticks(executions, JOB_CRON, BUCKET_TOLERANCE_MS)

    if not duplicates:
        log.info("No duplicate ticks across %d execution(s) for %s.", len(executions), JOB_WORKFLOW_ID)
        return

    for item in duplicates:
        log.warning(
            "Tick %s fired %d time(s): %s",
            item["tickBucket"], len(item["transactionIds"]), item["transactionIds"],
        )

    if not DRY_RUN:
        write_audit_report(JOB_WORKFLOW_ID, duplicates)

    log.info("Done. %d duplicate tick(s) %s.", len(duplicates), "to review" if DRY_RUN else "reported")


if __name__ == "__main__":
    run()

File: scheduled-job-runs-twice/node/find-duplicate-ticks.js
/**
 * Find Medusa v2 scheduled job ticks that fired more than once, because
 * more than one process is running in shared or worker WORKER_MODE against
 * the same database, with no distributed lock coordinating them.
 *
 * This is an infrastructure and config defect, not a data problem. It only
 * reports the duplicate ticks, it never resends a suppressed side effect
 * and never deletes a workflow_execution row. Safe to run again and again.
 *
 * Guide: https://www.allanninal.dev/medusa/scheduled-job-runs-twice/
 */
import { pathToFileURL } from "node:url";

const BASE_URL = process.env.MEDUSA_BACKEND_URL || "http://localhost:9000";
const EMAIL = process.env.MEDUSA_ADMIN_EMAIL || "admin@example.com";
const PASSWORD = process.env.MEDUSA_ADMIN_PASSWORD || "supersecret";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";
const JOB_WORKFLOW_ID = process.env.JOB_WORKFLOW_ID || "job-name";
const JOB_CRON = process.env.JOB_CRON || "*/15 * * * *";
const BUCKET_TOLERANCE_MS = Number(process.env.BUCKET_TOLERANCE_MS || 5000);

const EXECUTION_FIELDS = "id,transaction_id,workflow_id,created_at,state";

function parseField(field, lo, hi) {
  if (field === "*") return new Set(Array.from({ length: hi - lo + 1 }, (_, i) => lo + i));
  const values = new Set();
  for (const part of field.split(",")) {
    if (part.startsWith("*/")) {
      const step = Number(part.slice(2));
      for (let v = lo; v <= hi; v += step) values.add(v);
    } else if (part.includes("-")) {
      const [a, b] = part.split("-").map(Number);
      for (let v = a; v <= b; v++) values.add(v);
    } else {
      values.add(Number(part));
    }
  }
  return values;
}

function parseCron(cronExpression) {
  const [minute, hour, dom, month, dow] = cronExpression.trim().split(/\s+/);
  return {
    minute: parseField(minute, 0, 59),
    hour: parseField(hour, 0, 23),
    dom: parseField(dom, 1, 31),
    month: parseField(month, 1, 12),
    dow: parseField(dow, 0, 6),
  };
}

function fullRange(lo, hi) {
  return new Set(Array.from({ length: hi - lo + 1 }, (_, i) => lo + i));
}

function matches(date, spec) {
  if (!spec.minute.has(date.getUTCMinutes())) return false;
  if (!spec.hour.has(date.getUTCHours())) return false;
  if (!spec.month.has(date.getUTCMonth() + 1)) return false;
  const domOk = spec.dom.has(date.getUTCDate());
  const dowOk = spec.dow.has(date.getUTCDay());
  const domIsFull = spec.dom.size === fullRange(1, 31).size;
  const dowIsFull = spec.dow.size === fullRange(0, 6).size;
  if (!domIsFull && !dowIsFull) return domOk || dowOk;
  return domOk && dowOk;
}

export function nearestTickBoundary(cronExpression, at, searchMinutes = 1440) {
  // Minute-aligned tick boundary matching the cron spec closest to `at`.
  const spec = parseCron(cronExpression);
  const base = new Date(at.getTime());
  base.setUTCSeconds(0, 0);
  if (matches(base, spec)) return base;
  for (let offset = 1; offset <= searchMinutes; offset++) {
    const earlier = new Date(base.getTime() - offset * 60000);
    if (matches(earlier, spec)) return earlier;
    const later = new Date(base.getTime() + offset * 60000);
    if (matches(later, spec)) return later;
  }
  throw new Error("No matching tick boundary found within search window");
}

export function findDuplicateTicks(executions, cronSchedule, bucketToleranceMs = 5000) {
  // Pure: no I/O, no clock access. executions is a plain array of
  // { workflow_id, transaction_id, created_at } already fetched.
  const byWorkflow = new Map();
  for (const execution of executions) {
    const list = byWorkflow.get(execution.workflow_id) || [];
    list.push(execution);
    byWorkflow.set(execution.workflow_id, list);
  }

  const duplicates = [];
  for (const rows of byWorkflow.values()) {
    const buckets = new Map();
    for (const row of rows) {
      const createdAt = new Date(row.created_at);
      const tick = nearestTickBoundary(cronSchedule, createdAt);
      const deltaMs = Math.abs(createdAt.getTime() - tick.getTime());
      if (deltaMs > bucketToleranceMs) continue;
      const key = tick.toISOString();
      const set = buckets.get(key) || new Set();
      set.add(row.transaction_id);
      buckets.set(key, set);
    }

    for (const [tickBucket, txIds] of buckets.entries()) {
      if (txIds.size > 1) {
        duplicates.push({ tickBucket, transactionIds: [...txIds].sort() });
      }
    }
  }

  return duplicates.sort((a, b) => a.tickBucket.localeCompare(b.tickBucket));
}

async function getToken() {
  const res = await fetch(`${BASE_URL}/auth/user/emailpass`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ email: EMAIL, password: PASSWORD }),
  });
  if (!res.ok) throw new Error(`Medusa auth ${res.status}`);
  const body = await res.json();
  return body.token;
}

async function listWorkflowExecutions(token, workflowId, limit = 200) {
  const params = new URLSearchParams({
    workflow_id: workflowId, fields: EXECUTION_FIELDS, limit: String(limit), order: "created_at",
  });
  const res = await fetch(`${BASE_URL}/admin/workflows-executions?${params}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!res.ok) throw new Error(`Medusa ${res.status}`);
  const body = await res.json();
  return body.workflow_executions;
}

function writeAuditReport(jobName, duplicates) {
  // The only write this script does: an audit log line per duplicate tick.
  // Never resends a suppressed side effect, never deletes an execution row.
  for (const item of duplicates) {
    console.warn(
      `DUPLICATE TICK job=${jobName} tick=${item.tickBucket} transaction_ids=${JSON.stringify(item.transactionIds)} inferred_replicas=${item.transactionIds.length}`
    );
  }
}

export async function run() {
  const token = await getToken();
  const executions = await listWorkflowExecutions(token, JOB_WORKFLOW_ID);
  const duplicates = findDuplicateTicks(executions, JOB_CRON, BUCKET_TOLERANCE_MS);

  if (duplicates.length === 0) {
    console.log(`No duplicate ticks across ${executions.length} execution(s) for ${JOB_WORKFLOW_ID}.`);
    return;
  }

  for (const item of duplicates) {
    console.warn(`Tick ${item.tickBucket} fired ${item.transactionIds.length} time(s): ${JSON.stringify(item.transactionIds)}`);
  }

  if (!DRY_RUN) {
    writeAuditReport(JOB_WORKFLOW_ID, duplicates);
  }

  console.log(`Done. ${duplicates.length} duplicate tick(s) ${DRY_RUN ? "to review" : "reported"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
